Source/Scene/UrlTemplateImageryProvider.js

```diff
@@ -0,0 +1,601 @@
+     *                    this parameter is ignored and the tiling scheme's ellipsoid is used instead. If neither
+     *                    parameter is specified, the WGS84 ellipsoid is used.
+     * @param {Number} [options.tileWidth=256] Pixel width of image tiles.
+     * @param {Number} [options.tileHeight=256] Pixel height of image tiles.
+     *
+     * @see ArcGisMapServerImageryProvider
+     * @see BingMapsImageryProvider
+     * @see GoogleEarthImageryProvider
+     * @see OpenStreetMapImageryProvider
+     * @see WebMapTileServiceImageryProvider
+     * @see SingleTileImageryProvider
+     * @see WebMapServiceImageryProvider
+     *
+     * @example
+     * // A tile provider with z/x/y, .png tile files, and tile Y coordinates numbered from the South (TMS style).
+     * var utip = new Cesium.UrlTemplateImageryProvider({
+     *     url : '../images/cesium_maptiler/Cesium_Logo_Color/{z}/{x}/{reverseY}.png',
+     *     maximumLevel: 4,
+     *     rectangle: new Cesium.Rectangle.fromDegrees(-120, 20, -60, 40)
+     * });
+     * // An emulation of a WMS server with a time dimension
+     * utip = new Cesium.UrlTemplateImageryProvider({
+     *    url : 'URL/to/WMS?SERVICE=WMS&VERSION=1.1.1&REQUEST=GetMap&LAYERS=980_13'+
+     *    '&FORMAT=image/png&STYLES=default&TRANSPARENT=true&SRS=EPSG:4326&'+
+     *    'BBOX={westDegrees},{southDegrees},{eastDegrees},{northDegrees}&WIDTH=256&HEIGHT=256&TIME=2009-11-30T12:00',
+     * });
+     * // An emulation of a WMTS server 
+     * utip = new Cesium.UrlTemplateImageryProvider({
+     *    url : 'URL/to/WMTS?REQUEST=gettile&LAYER=980_13&FORMAT=image/png&TILEMATRIXSET=EPSG:4258'+
+     *    '&TILEMATRIX=EPSG:4258:{Z}&TILEROW={Y}&TILECOL={X}&DIM_TIME=2013-11-20T11:15:00Z&STYLE=default
+     * });
+     */
+    var UrlTemplateImageryProvider = function UrlTemplateImageryProvider(options) {
+        //>>includeStart('debug', pragmas.debug);
+        if (!defined(options) || !defined(options.url)) {
+            throw new DeveloperError('options.url is required.');
+        }
+        //>>includeEnd('debug');
+
+        this._url = options.url;
+        this._proxy = options.proxy;
+        this._tileDiscardPolicy = options.tileDiscardPolicy;
+        this._errorEvent = new Event();
+        
+        this._subdomains = options.subdomains;
+        if (Array.isArray(this._subdomains)) {
+            this._subdomains = this._subdomains.slice();
+        } else if (defined(this._subdomains) && this._subdomains.length > 0) {
+            this._subdomains = this._subdomains.split('');
+        } else {
+            this._subdomains = ['a', 'b', 'c'];
+        }
+
+        this._tileWidth = defaultValue(options.tileWidth, 256);
+        this._tileHeight = defaultValue(options.tileHeight, 256);
+        this._minimumLevel = defaultValue(options.minimumLevel, 0);
+        this._maximumLevel = defaultValue(options.maximumLevel, 18);
```

Comment:
It's fine if this is outside the scope of this PR, but are we sure we want to keep this default as 18?  We have ran into issues where users thought Cesium couldn't handle their high-res imagery because of this.  Are we paying a big cost by increasing it?